yaptide/celery/utils/utils.py

```diff
@@ -144,26 +149,36 @@ def read_file(filepath: Path, simulation_id: int, task_id: str, update_key: str)
         send_task_update(simulation_id, task_id, update_key, up_dict)
 
     loglines = log_generator(logfile)
+    requested_primaries = 0
     for line in loglines:
         utc_now = datetime.utcnow()
         if re.search(RUN_MATCH, line):
-            if utc_now.timestamp() - update_time < 2:  # hardcoded 2 seconds to avoid spamming
-                continue
-            update_time = utc_now.timestamp()
             splitted = line.split()
+            simulated_primaries = int(splitted[3])
             up_dict = {
-                "simulated_primaries": int(splitted[3]),
+                "simulated_primaries": simulated_primaries,
                 "estimated_time": int(splitted[9])
                 + int(splitted[7]) * 60
                 + int(splitted[5]) * 3600
             }
-            send_task_update(simulation_id, task_id, update_key, up_dict)
+            if utc_now.timestamp() > update_time + seconds_before_next_update:  # time to wait before next update
+                send_task_update(simulation_id, task_id, update_key, up_dict)
+                update_time = utc_now.timestamp()
+            # block of code below makes the updates not so frequent
+            # updates comes every `seconds_before_next_update` seconds
+            # this is convenient for long simulations, but we may also miss this way some updates,
+            # like the one when simulation is finished (simulated_primaries == requested_primaries)
+            # therefore we send update when simulation is finished
+            if simulated_primaries == requested_primaries:
+                send_task_update(simulation_id, task_id, update_key, up_dict)
+                update_time = utc_now.timestamp()
```

Comment:
Blocks of code after both if statements are equal. Furthermore it is possible for both statements to be true, so task will be updated 2 times in the same iteration. Suggestions:
- merge statements with `or`
- or at the end of first if add `continue`